tempodb/backend/block_meta.go

```diff
@@ -73,3 +179,99 @@ func (b *BlockMeta) ObjectAdded(id []byte, start uint32, end uint32) {
 
 	b.TotalObjects++
 }
+
+func (b *BlockMeta) DedicatedColumnsHash() uint64 {
+	return b.DedicatedColumns.Hash()
+}
+
+func DedicatedColumnsFromTempopb(tempopbCols []*tempopb.DedicatedColumn) (DedicatedColumns, error) {
+	cols := make(DedicatedColumns, 0, len(tempopbCols))
+
+	for _, c := range tempopbCols {
+		scope, err := DedicatedColumnScopeFromTempopb(c.Scope)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to convert dedicated column '%s'", c.Name)
+		}
+
+		typ, err := DedicatedColumnTypeFromTempopb(c.Type)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to convert dedicated column '%s'", c.Name)
+		}
+
+		cols = append(cols, DedicatedColumn{
+			Scope: scope,
+			Name:  c.Name,
+			Type:  typ,
+		})
+	}
+
+	return cols, nil
+}
+
+func (dcs DedicatedColumns) ToTempopb() ([]*tempopb.DedicatedColumn, error) {
+	tempopbCols := make([]*tempopb.DedicatedColumn, 0, len(dcs))
+
+	for _, c := range dcs {
+		scope, err := c.Scope.ToTempopb()
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to convert dedicated column '%s'", c.Name)
+		}
+
+		typ, err := c.Type.ToTempopb()
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to convert dedicated column '%s'", c.Name)
+		}
+
+		tempopbCols = append(tempopbCols, &tempopb.DedicatedColumn{
+			Scope: scope,
+			Name:  c.Name,
+			Type:  typ,
+		})
+	}
+
+	return tempopbCols, nil
+}
+
+func (dcs DedicatedColumns) Validate() error {
+	for _, dc := range dcs {
+		err := dc.Validate()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (dc *DedicatedColumn) Validate() error {
+	if dc.Name == "" {
```

Comment:
For reading and writing the well known attributes are always handled first. This means that it has no effect - except one spare column being wasted that will remain empty.

It would be good to validate this. However, to do this properly we would need to make well known attributes available in `tempodb/backend`. To do this we would have to either import something from `encoding/vparquet3` directly or extend `VersionedEncoding`. Given the small impact of well known attributes in the dedicated column configuration, I think it’s OK not to validate this.